src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -74,36 +91,58 @@ void AsyncLogWriter::enqueue_locked(LogFileStreamOutput* output, const LogDecora
   // client should use "" instead.
   assert(msg != nullptr, "enqueuing a null message!");
 
-  if (!_buffer->push_back(output, decorations, msg)) {
+  size_t msg_len = strlen(msg);
+
+  ConsumerLocker clocker;
+  if (_buffer->push_back(output, decorations, msg, msg_len)) {
+    _data_available = true;
+    clocker.notify();
+    return;
+  }
+
+  if (LogConfiguration::async_mode() == LogConfiguration::AsyncMode::Stall) {
+    size_t size = Message::calc_size(msg_len);
+    void* ptr = os::malloc(size, mtLogging);
+    if (ptr == nullptr) {
+      // Out of memory. We bail without any notice.
+      // Some other part of the system will probably fail later.
+      return;
+    }
+    new (ptr) Message(output, decorations, msg, msg_len);
+    _stalled_message = (Message*)ptr;
+    while (_stalled_message != nullptr) {
+      clocker.wait();
```

Comment:
What ensures that the consumer side is notified and will process the stalled_message if the push_back above failed on an empty buffer?

Reply:
I'm missing something here. If the message is too big even when the buffer is empty then how do we proceed? Stalling to wait for more space is not going to work. I assumed the message is already getting broken up into chunks somewhere else. ??

Reply:
@dholmes-ora, We proceed by explicitly malloc:ing the amount of memory necessary for the message. We never stall to wait for more space, we stall because we've put something into _stalled_message and wait for it to be flushed.